Add doc for observability (#673)
smallrye/smallrye-stork: docs/docs/observability.md

# Stork Observability API

Stork is observable. 
Stork proposes an observability API that automatically observes some parameters in order to show how the Stork service discovery and selection are behaving.

For any Observation to happen, you need to provide your own implementation of a `ObservationCollector`. By default, Stork provides a no-op implementation.
The `ObservationCollector` is responsible for instantiating the `StorkObservation`. 
The `StorkObservation` reacts to Stork events thanks to a `StorkEventHandler`. You can extend the metrics collection by extending the `StorkEventHandler` interface.

The following sequence diagram shows how the observability is initialized : 

![observability initialization](target/observability_sequence.svg#only-light)
![observability initialization](target/observability_sequence_dark.svg#only-dark)

> The `StorkObservation` registers times, discovered instances, the selected instance and failures by reacting to the lifecycle of a Stork event such us:

Author: mmm...yes, I aggree. We can remove it and add it if we feel the need

> - start : Observation has been started. Begin time is registered. Happens when the `ObservationCollector#create()` method gets called.

Reviewer: Why didn't you integrate this change?

> - service discovery success: a collection of instances has been successfully discovered for a service. End discovery time and number of instances are registered. Happens when the `StorkObservation#onServiceDiscoverySuccess` gets called.

Reviewer: Why didn't you integrate this change?

> - service discovery error: an error occurred when discovering service. End discovery time and failure cause are registered. Happens when the `StorkObservation#onServiceDiscoveryFailure` gets called.

Reviewer: Why didn't you integrate this change?

> - service selection success: an instance has been successfully selected from the collection. End selection time and selected instance id are registered. Happens when the `StorkObservation#onServiceSelectionSuccess` gets called.

Reviewer: Why didn't you integrate this change?

> - service selection error: an error occurred during selecting the instance. End selection time and failure cause are registered. Happens when the `StorkObservation#onServiceSelectionFailure` gets called.

Reviewer: Why didn't you integrate this change?

> - end: Observation has finished. Overall duration is registered. Happens when the `StorkObservation#onServiceSelectionSuccess` gets called.

Reviewer: Why didn't you integrate this change?

The following sequence diagram represents the described observation process above:

  ![observation_process](target/observation_sequence.svg#only-light)
  ![observation_process](target/observation_sequence_dark.svg#only-dark)

## Implementing an observation collector

> An `ObservationCollector` implementation needs to override the `create` method in order to provide an instance of StorkObservation.

Reviewer: ```suggestion
An `ObservationCollector` implementation must override the `create` method to provide an instance of StorkObservation.
```

Reviewer: Why didn't you integrate this change?

A custom observation collector class should look as follows:

```java linenums="1"
{{ insert('examples/AcmeObservationCollector.java') }}
```

> The next step is to initialize Stork with an `ObservableStorkInfrastructure` taking an instance of your ObservationCollector as parameter.

Reviewer: ```suggestion
The next step is to initialize Stork with an `ObservableStorkInfrastructure`, taking an instance of your `ObservationCollector` as parameter.
```

```java linenums="1"
{{ insert('examples/ObservableInitializationExample.java') }}
```

> Then, Stork will use your implementation to register metrics.

Reviewer: ```suggestion
Then, Stork uses your implementation to register metrics.
```

## Observing service discovery and selection behaviours

To access metrics registered by `StorkObservation`, use the following code:

```java linenums="1"
{{ insert('examples/ObservationExample.java') }}
```

> 

Reviewer: Too many empty lines.

# Stork Observability with Quarkus

> Stork metrics are automatically enabled when using Stork together with Micrometer extension in a Quarkus application. Micrometer collects the metrics of the Rest or Grpc client using Stork.

Reviewer: ```suggestion
Stork metrics are automatically enabled when using Stork together with the Micrometer extension in a Quarkus application. 
Micrometer collects the metrics of the rest clients and grpc clients using Stork, as well as when using the Stork API.
```
